agent/taskresource/credentialspec/credentialspec_linux.go

```diff
@@ -0,0 +1,404 @@
+
+	if !strings.HasPrefix(credSpecFile, "file://") {
+		err := errors.New("invalid credentialspec file specification")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	fileName := strings.SplitAfterN(credSpecFile, "file://", 2)
+	data, err := ioutil.ReadFile(fileName[1])
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	credSpecData := string(data)
+
+	cs.updateCredSpecMapping(credentialSpec, credSpecData)
+}
+
+func (cs *CredentialSpecResource) handleS3CredentialspecFile(originalCredentialSpec, credentialSpecS3ARN string, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+	if iamCredentials == (credentials.IAMRoleCredentials{}) {
+		err := errors.New("credentialspec resource: unable to find execution role credentials")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	_, err := arn.Parse(credentialSpecS3ARN)
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	bucket, key, err := s3.ParseS3ARN(credentialSpecS3ARN)
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	s3Client := cs.s3ClientCreator.NewS3Client(cs.region, iamCredentials)
+
+	credSpecJsonStringUnformatted, err := s3.GetObject(bucket, key, s3Client)
+
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	credSpecJsonStringBytes := &bytes.Buffer{}
+	json.Compact(credSpecJsonStringBytes, []byte(credSpecJsonStringUnformatted))
+	credSpecJsonString := credSpecJsonStringBytes.String()
+
+	cs.updateCredSpecMapping(originalCredentialSpec, credSpecJsonString)
+}
+
+func (cs *CredentialSpecResource) handleSSMCredentialspecFile(originalCredentialSpec, credentialSpecSSMARN string, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	if iamCredentials == (credentials.IAMRoleCredentials{}) {
+		err := errors.New("credentialspec resource: unable to find execution role credentials")
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	parsedARN, err := arn.Parse(credentialSpecSSMARN)
+	if err != nil {
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+
+	// An SSM ARN is in the form of arn:aws:ssm:us-west-2:123456789012:parameter/a/b. The parsed ARN value
+	// would be parameter/a/b. The following code gets the SSM parameter by passing "/a/b" value to the
+	// GetParametersFromSSM method to retrieve the value in the parameter.
+	ssmParam := strings.SplitAfterN(parsedARN.Resource, "parameter", 2)
+	if len(ssmParam) != 2 {
+		err := fmt.Errorf("the provided SSM parameter:%s is in an invalid format", parsedARN.Resource)
+		cs.setTerminalReason(err.Error())
+		errorEvents <- err
+		return
+	}
+	ssmParams := []string{ssmParam[1]}
+
+	ssmClient := cs.ssmClientCreator.NewSSMClient(cs.region, iamCredentials)
```

Comment:
can you use the package's ssm client directly (see s3 examples above).  This will also allow you to test using the core mocks.

Reply:
we can have a chat, about it. I am using factory clients already